exercises/say/.meta/src/reference/java/Say.java

```diff
@@ -0,0 +1,204 @@
+import java.util.Arrays;
+
+public class Say {
+
+    public String say(long number) {
+        if (number < 0) {
+            throw new IllegalArgumentException("input out of range");
+        }
+        char[] digits = String.valueOf(number).toCharArray();
+        if (digits.length > 12) {
+            throw new IllegalArgumentException("input out of range");
+        }
+        char[][] processedDigits = process(digits);
+
+        String message = "";
+        for (int index = 0; index < processedDigits.length; index++) {
+            char[] blockDigits = processedDigits[index];
+            message = concatWithSpace(message,
+                    sayDigits(blockDigits, processedDigits.length - index));
+        }
+        return message;
+    }
+
+    private String sayDigits(char[] digits, int orderNumber) {
+        String sentence = "";
+
+        if (digits.length == 3) {
+            digits = filterInitialZeros(digits);
```

Comment:
Annoyingly, there is a conflict here with the track policies:

> Avoid using final in user-facing code. Liberal use of the `final` keyword remains controversial. It adds little value to test suites, and if included in starter implementations, places implicit restrictions on user solutions.

My work has more-or-less the same policy (only using the final keyword for class constants and constructor assigned member variables). Even at that, we still look at it as a bad practice to re-assign a parameter because you can lose track of the reassignment. Someone comes along later to edit the same code and does not expect the parameter to have changed so they try to check something about the input and then they debug for a while until they discover that somewhere else the input parameter has been reassigned. Principle of least surprise says don't do it.